.github/workflows/ci.yml

```diff
@@ -12,7 +12,15 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        ghc: ["8.10.1", "8.10.2", "8.10.4", "8.10.7"]
+        ghc:
+          # NB: PyF doesn't work with GHC <8.4 (guibou/PyF#94)
+          - "8.4.1"
+          - "8.6.1"
+          - "8.8.1"
```

Comment:
We should test on the earliest, because that _should_ imply it works on the later revisions. If it doesn’t, that’s an upstream bug. Of course, that happens, which is why we also test on 8.10.7. If we only test on the later revisions, it tells us nothing about the earlier ones.

Reply:
Does a higher version not also imply a lower version? I thought the 3rd component of the version is only for minor bug fixes. I suggested testing the highest version just because I'm guessing most people are on the highest version.

Anyway, since there are at least two instances where the GHC API doesn't respect that, maybe we should just test all versions - after all it's free.

Reply:
Even though GHC only has a 3-component version, in practice it follows the Haskell PVP where the first two components are for breaking changes (otherwise we could test just 8.0.1 and 9.0.1).

I'm not opposed to testing all of them, but GH does limit the number of concurrent jobs to 20. I don't know what it would be, but if we added another dimension to our matrix, we'd be at 16 right now, so I was trying to minimize it to avoid longer builds. I also wanted to start testing against ghc-HEAD with failure allowed, so we can be made aware when upstream changes get in our way.

Reply:
I wrote a reply to this yesterday, but I don't know what happened to it, so I'll try to recreate it ...

GHC only has a 3-component version, but it still follows the Haskell PVP -- i.e., the first _two_ components are breaking changes, the third is non-breaking, but may include new features. So testing on a later version would allow us to accidentally start relying on added features without catching that we no longer work on older versions.

When upstream breaks those expectations, like 8.10.7 (and apparently 9.2.2), then we should test. In the case of 8.10.7, we explicitly run into the incompatibilities and have to conditionalize, so that's a clear case we have to test, since that code only runs on that version. For 9.2.2, I don't _think_ that's the case, so it's less clear-cut (can you point me to the breaking changes there?).

It's free to have a bunch of jobs, but they stop being concurrent at 20. We only have 9 at the moment, so we have some headroom, but I'd at least want to tag the lower-priority jobs in some way (like a comment) so that if we do add another dimension to the matrix or something, we can trim those jobs to keep the build time down.

I also want to start testing against ghc-HEAD (with failures allowed to pass) so we can be made aware of breaking issues ASAP.

Reply:
9.2.2 adds another strict field to `SimplMode`: `sm_cast_swizzle`. You can try building this branch: https://github.com/con-kitty/categorifier/tree/ghc922

Reply:
> So testing on a later version would allow us to accidentally start relying on added features without catching that we no longer work on older versions.

I see. Makes sense.

> I'd at least want to tag the lower-priority jobs in some way (like a comment)

Yeah eventually this is a good thing to do. I'm not sure GHC API respects any versioning semantics :joy:, so it's useful to test them all, but for now I think this is fine.

Reply:
I added another commit, with GHC 9.2.2 support.